### PR TITLE
More themable selectInput()

### PR DESCRIPTION
**Pull request by cpsievert**
This PR allows `selectInput()` to be styled via {bootstraplib}  (see also https://github.com/rstudio/shiny/pull/2964 and https://github.com/rstudio/shiny/pull/2958)

## Testing notes

Install `remotes::install_github("rstudio/shiny")`. Now if you run the following app you should see a dropdown with the same colors/fonts as the screenshot. Now change `bs_theme_new()` to `bs_theme_new(3)` and re-run the app. You should see a similar result.

```r
library(shiny)
library(bootstraplib)
shinyOptions(bootstraplib = TRUE)

bs_theme_new()
bs_theme_base_colors(bg = "#002B36", fg = "#EEE8D5")
bs_theme_accent_colors(primary = "#2AA198")

ui <- fluidPage(
  tags$h4("With selectize"),
  selectInput("single", "Choose single", choices = state.abb),
  selectInput("multiple", "Choose multiple", choices = state.abb, multiple = TRUE),
  tags$h4("Without selectize"),
  selectInput("single2", "Choose single", choices = state.abb, selectize = FALSE),
  selectInput("multiple2", "Choose multiple", choices = state.abb, multiple = TRUE, selectize = FALSE),
)

shinyApp(ui, function(input, output) {})
```

Screen Shot 2020-08-17 at 1 52 29 PM

### CODE REVIEW

File: R/bootstrap.R [hunk 1]
```diff
@@ -42,10 +42,20 @@ bootstrapPage <- function(..., title = NULL, responsive = NULL, theme = NULL) {
       # remainder of tags passed to the function
       list(...)
     ),
-    bootstrapLib()
+    if (useBsTheme()) bootstraplib::bootstrap() else bootstrapLib()
   )
 }
 
+useBsTheme <- function() {
+  if (!isTRUE(getOption("shiny.bootstraplib"))) {
```
Comment (cpsievert) [inline R/bootstrap.R:50]:
```suggestion
  if (!isTRUE(getShinyOption("shiny.bootstraplib"))) {
```

File: R/input-select.R [hunk 2]
```diff
@@ -185,16 +185,58 @@ selectizeInput <- function(inputId, ..., options = NULL, width = NULL) {
...
-    "selectize", "0.11.2", c(href = "shared/selectize"),
-    stylesheet = "css/selectize.bootstrap3.css",
-    head = format(tagList(
-      HTML('<!--[if lt IE 9]>'),
-      tags$script(src = 'shared/selectize/js/es5-shim.min.js'),
-      HTML('<![endif]-->'),
-      tags$script(src = 'shared/selectize/js/selectize.min.js')
-    ))
-  )
+  # Compile selectize Sass against a bootstraplib theme (if one exists)
```
Comment (cpsievert) [inline R/input-select.R:188]: Does this work with real-time theming?

File: R/shiny-options.R [hunk 3]
```diff
@@ -51,6 +51,10 @@ getShinyOption <- function(name, default = NULL) {
 #'   The default polling interval is 500 milliseconds. You can change this
 #'   by setting e.g. `options(shiny.autoreload.interval = 2000)` (every
 #'   two seconds).}
+#' \item{shiny.bootstraplib (defaults to `FALSE`)}{When a global **bootstraplib** theme
```
Comment (cpsievert) [inline R/shiny-options.R:54]:
```suggestion
#' \item{bootstraplib (defaults to `FALSE`)}{When a global **bootstraplib** theme
```

File: inst/www/shared/selectize/scss/selectize.default.scss [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+@import "selectize";
+
+$selectize-color-item: #1da7ee;
+$selectize-color-item-text: #fff;
+$selectize-color-item-active-text: #fff;
+$selectize-color-item-border: #0073bb;
+$selectize-color-item-active: #92c836;
+$selectize-color-item-active-border: #00578d;
+$selectize-width-item-border: 1px;
+$selectize-caret-margin: 0 1px;
```
Comment (cpsievert) [inline inst/www/shared/selectize/scss/selectize.default.scss:26]: Perhaps we could add an argument to `selectInput()` to set these values (in a similar way to the upcoming `sliderInput()` PR)

File: R/input-select.R [hunk 5]
```diff
@@ -185,16 +185,58 @@ selectizeInput <- function(inputId, ..., options = NULL, width = NULL) {
...
+            "selectize-color-item-active-text" = "$white !default;",
+            "selectize-color-item-active-border" = "$gray-base !default;"
+          )
+        } else {
+          list(
+            "selectize-color-item" = "gray('100') !default;",
+            "selectize-color-item-border" = "gray('400') !default;",
+            "selectize-color-item-active-text" = "$white !default;",
+            "selectize-color-item-active-border" = "$black !default;"
+          )
```
Comment (cpsievert) [inline R/input-select.R:217]: For some reason this isn't enough to get Bootswatch based dropdowns colored appropriately:

```r
library(bootstraplib)
bs_theme_new(bootswatch = "darkly")
bs_theme_preview()
```

Comment (cpsievert) [reply R/input-select.R]: There are two problems here:

1. Bootswatch themes don't take the same approach as `bs_theme_base_colors()` (the latter sets the "base" colorscales, which effects all downstream color variables). I'm not sure whether it'll make sense for use to "patch" this in the sense that bootswatch color model matches `bs_theme_base_colors()`, but assuming we don't, this implies we can't use `gray()` at this point in the way I was hoping to.
2. Bootswatch themes like darkly and cyborg have two totally different color models for inputs (light bg, dark fg) and dropdowns (dark bg, light fg)....

File: inst/www/shared/selectize/scss/selectize.bootstrap4.scss [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+$selectize-arrow-offset: calc(#{$selectize-padding-x} + 5px) !default;
+
+
+@import "selectize";
+
+//Import Plugins
+@import "plugins/drag_drop";
+@import "plugins/dropdown_header";
+@import "plugins/optgroup_columns";
+@import "plugins/remove_button";
```
Comment (cpsievert) [inline inst/www/shared/selectize/scss/selectize.bootstrap4.scss:63]: Do these plugins need to be imported in the BS3 case?

Comment (cpsievert) [reply inst/www/shared/selectize/scss/selectize.bootstrap4.scss:63]: Yep, done in 77e9969be07d161fe7a428116f3eabfb647d35d9

File: tools/updateSelectize.R [hunk 7]
```diff
@@ -56,20 +50,66 @@ file.copy(
...
+dir.create(target)
+file.rename(
+  file.path(tmpdir, sprintf("selectize-bootstrap-4-style-%s/src/selectize", version)),
+  target
+)
+
+# Remove the unnecessary imports of Bootstrap
+scss_file <- "inst/www/shared/selectize/scss/selectize.bootstrap4.scss"
+scss <- readLines(scss_file)
+scss <- scss[!grepl('@import\\s+"\\.\\./bootstrap', scss)]
```
Comment (cpsievert) [inline tools/updateSelectize.R:77]: Maybe this should be done in the patch now instead

File: R/bootstrap.R [hunk 8]
```diff
@@ -42,10 +42,37 @@ bootstrapPage <- function(..., title = NULL, responsive = NULL, theme = NULL) {
...
+    stop("Shiny's bootstraplib option requires a bootstraplib theme to be active. Initialize one with `bootstraplib::bs_theme_new()`.", call. = FALSE)
+  }
+  TRUE
+}
+
+# Reusable function for input widgets to compile their Sass against a bootstraplib theme
+bootstrapSass <- function(sassInput, pattern = "file", ...) {
+  # TODO: outFile should eventually become something like
+  # file.path(tempdir(), paste(pattern, "-", sass::hash(sassInput), ".min.css"))
+  outFile <- tempfile(pattern = pattern, fileext = ".min.css")
```
Comment (cpsievert) [inline R/bootstrap.R:66]:
```suggestion
  outFile <- file.path(tempdir(), paste(pattern, "-", sass::sass_hash(sassInput), ".min.css"))
```

File: DESCRIPTION [hunk 9]
```diff
@@ -8,6 +8,7 @@ Authors@R: c(
     person("JJ", "Allaire", role = "aut", email = "jj@rstudio.com"),
     person("Yihui", "Xie", role = "aut", email = "yihui@rstudio.com"),
     person("Jonathan", "McPherson", role = "aut", email = "jonathan@rstudio.com"),
+    person("Carson", "Sievert", role = "aut", email = "carson@rstudio.com"),
```
Comment (wch) [inline DESCRIPTION:11]: This is overdue!

**Review by wch — APPROVED** [review]

**Pull request merged**